packages/cozy-konnector-libs/src/libs/saveBills.js

```diff
@@ -197,4 +203,68 @@ function checkRequiredAttributes(entries) {
     }
   }
 }
+
+function appendContractsDatas(tempEntries, options) {
+  if (options.contractLabel && options.contractId === undefined) {
+    log('warn', 'contractLabel used without contractId')
+    return tempEntries
+  }
+  let newEntries = tempEntries
+  if (options.contractId) {
+    // if contractId passed by option
+    newEntries = appendContractDatasByOption(newEntries, options)
+  } else if (billsHaveContractId(newEntries)) {
+    // if contractId passed by bill attribute
+    newEntries = appendContractDatasByAttribute(newEntries, options)
+  }
+  return newEntries
+}
+
+function appendContractDatasByOption(newEntries, options) {
+  // Surcharge label if not define
+  if (!options.contractLabel) {
+    options.contractLabel = options.contractId
+  }
+  options.subPath = options.contractLabel
+
+  // Add contractId to deduplication keys
+  if (options.keys) {
+    options.keys.push('contractId')
+  }
+
+  // Add datas in bills
+  for (let entry of newEntries) {
+    entry.contractLabel = options.contractLabel
+    entry.contractId = options.contractId
+  }
+  return newEntries
+}
+
+function appendContractDatasByAttribute(newEntries, options) {
```

Comment:
The append verb is mainly used when pushing an element in a list, not when merging data inside items of a list.